drafts/compound-components.md

```diff
@@ -0,0 +1,252 @@
+---
+title: The Art of Component Composition - Leveraging Compound Components in React
+domain: software-engineering-corner.zuehlke.com
+tags: reactjs, components, patterns
+cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1709986147750/6D9yz4Rsi.png?auto=format
+publishAs: gadu
+hideFromHashnodeCommunity: false
+saveAsDraft: true
+---
+
+As frontend applications get bigger and bigger, they have to cover more and more use cases and therefore their UI components need to get more flexible. This blog post describes how the requirements for a UI component increase throughout a typical project. We show what problems this entails and how they can be solved with the compound component pattern.
+
+## Exploring the issue by the example of a dialog
+
+We have a React application that can manage files. One feature is to delete files, another is to duplicate them. For both actions, the action should be confirmed with a dialog. Since deleting is a destructive action, we also want to color the delete button red. In such a case, we want to implement a dialog component that we can use for both scenarios. We, therefore, need various parameters to configure the dialog accordingly. The following code shows how the call of such a component could look like:
+
+```typescript
+<Dialog
+  title="Delete File"
+  content="Do you really want to delete this file? You cannot undo this action."
+  labelActionButton="Delete"
+  isDestructiveAction
+/>
+```
+
+The dialog now looks like this:
+
+![Delete Dialog](https://cdn.hashnode.com/res/hashnode/image/upload/v1709470151176/qBBC62QhH.png?auto=format)
+
+The next requirement is to add a renaming dialog, where you can enter the new file name. The content of the dialog does not only consist of text but also contains an input field. We must be able to pass this to the dialog.
+Another requirement is to convert documents into images. The confirmation dialog should now have three buttons: "Cancel", "Convert to PNG" and "Convert to JPG". We now need to introduce a third button and make it configurable.
+We are only at the beginning of our application and the number of parameters is already starting to get out of hand and there are still many requirements to come.
+
+If we don't want a huge, confusing dialog component that is full of configuration parameters and conditions, we need to take a different approach. One approach that is often chosen is to implement different subtypes of dialogs. Instead of one dialog, you have multiple dialogs for different use cases. One example could be a ConfirmDialog, a dialog that is exclusively used for confirmations of actions.
```

Comment:
i am bad a punctuation, but for me it makes more sense to have a comma there :D 
```suggestion
If we don't want a huge, confusing dialog component that is full of configuration parameters and conditions, we need to take a different approach. One approach that is often chosen, is to implement different subtypes of dialogs. Instead of one dialog, you have multiple dialogs for different use cases. One example could be a ConfirmDialog, a dialog that is exclusively used for confirmations of actions.
```